src/qt/bantablemodel.cpp

```diff
@@ -0,0 +1,192 @@
+
+#include "clientmodel.h"
+#include "guiconstants.h"
+#include "guiutil.h"
+
+#include "net.h"
+#include "sync.h"
+#include "utiltime.h"
+
+#include <QDebug>
+#include <QList>
+#include <QTimer>
+
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <boost/date_time/c_local_time_adjustor.hpp>
+
+// private implementation
+class BanTablePriv
+{
+public:
+    /** Local cache of peer information */
+    QList<CCombinedBan> cachedBanlist;
+    /** Column to sort nodes by */
+    int sortColumn;
+    /** Order (ascending or descending) to sort nodes by */
+    Qt::SortOrder sortOrder;
+
+    /** Pull a full list of banned nodes from CNode into our cache */
+    void refreshBanlist()
+    {
+        std::map<CSubNet, int64_t> banMap;
+        CNode::GetBanned(banMap);
+
+        cachedBanlist.clear();
+#if QT_VERSION >= 0x040700
+        cachedBanlist.reserve(banMap.size());
+#endif
+        std::map<CSubNet, int64_t>::iterator iter;
+        for (iter = banMap.begin(); iter != banMap.end(); ++iter) {
+            CCombinedBan banEntry;
+            banEntry.subnet = iter->first;
+            banEntry.bantil = iter->second;
+            cachedBanlist.append(banEntry);
+        }
+    }
+
+    int size()
+    {
+        return cachedBanlist.size();
+    }
+
+    CCombinedBan *index(int idx)
+    {
+        if(idx >= 0 && idx < cachedBanlist.size()) {
+            return &cachedBanlist[idx];
+        } else {
+            return 0;
+        }
+    }
+};
+
+BanTableModel::BanTableModel(ClientModel *parent) :
+    QAbstractTableModel(parent),
+    clientModel(parent),
+    timer(0)
+{
+    columns << tr("IP/Netmask") << tr("Banned Until");
+    priv = new BanTablePriv();
+    // default to unsorted
+    priv->sortColumn = -1;
+
+    // set up timer for auto refresh
+    timer = new QTimer();
+    connect(timer, SIGNAL(timeout()), SLOT(refresh()));
+    timer->setInterval(MODEL_UPDATE_DELAY);
+
+    // load initial data
+    refresh();
+}
+
+void BanTableModel::startAutoRefresh()
+{
+    timer->start();
+}
+
+void BanTableModel::stopAutoRefresh()
+{
+    timer->stop();
+}
+
+int BanTableModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent);
+    return priv->size();
+}
+
+int BanTableModel::columnCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent);
+    return columns.length();;
+}
+
+QVariant BanTableModel::data(const QModelIndex &index, int role) const
+{
+    if(!index.isValid())
+        return QVariant();
+
+    CCombinedBan *rec = static_cast<CCombinedBan*>(index.internalPointer());
+
+    if (role == Qt::DisplayRole) {
+        switch(index.column())
+        {
+        case Address:
+            return QString::fromStdString(rec->subnet.ToString());
+        case Bantime:
+            //show time in users local timezone, not 64bit compatible!
+            //TODO find a way to support 64bit timestamps
+            boost::posix_time::ptime pt1 = boost::posix_time::from_time_t(rec->bantil);
```

Comment:
I think this won't work either. The [docs](http://doc.qt.io/qt-5.4/qdatetime.html#fromTime_t) are saying: `QDateTime QDateTime::fromTime_t(uint seconds)` (uint!).